File: src/main/java/com/github/zxh/classpy/classfile/attribute/Code.java
package com.github.zxh.classpy.classfile.attribute;

import com.github.zxh.classpy.classfile.ClassComponent;
import com.github.zxh.classpy.classfile.bytecode.InstructionFactory;
import com.github.zxh.classpy.classfile.constant.ConstantPool;
import com.github.zxh.classpy.classfile.datatype.U4;
import com.github.zxh.classpy.classfile.reader.ClassReader;
import com.github.zxh.classpy.classfile.bytecode.Instruction;
import com.github.zxh.classpy.classfile.bytecode.Opcode;

import java.util.List;

public class Code extends ClassComponent {

    private final U4 codeLength;

    public Code(U4 codeLength) {
        this.codeLength = codeLength;
    }
    
    @Override
    protected void readContent(ClassReader reader) {
        final int startPosition = reader.getPosition();
        final int endPosition = startPosition + codeLength.getValue();
        
        int position;
        while ((position = reader.getPosition()) < endPosition) {
            int pc = position - startPosition;
            byte b = reader.getByteBuffer().get(position);
            Opcode opcode = Opcode.valueOf(Byte.toUnsignedInt(b));
            Instruction instruction = InstructionFactory.create(opcode, pc);
            instruction.read(reader);
            add(instruction);
        }
    }

    @Override
    protected void afterRead(ConstantPool cp) {
        List<ClassComponent> instructions = super.getSubComponents();

        int maxPc = ((Instruction) instructions.get(instructions.size() - 1)).getPc();
        int pcWidth = String.valueOf(maxPc).length();
        String fmtStr = "%0" + pcWidth + "d";
        for (ClassComponent c : instructions) {
            Instruction instruction = (Instruction) c;
            instruction.setName(String.format(fmtStr, instruction.getPc()));
        }
    }

}

File: src/main/kotlin/com/github/zxh/classpy/classfile/attribute/Code.kt
package com.github.zxh.classpy.classfile.attribute;

import com.github.zxh.classpy.classfile.ClassComponent;
import com.github.zxh.classpy.classfile.bytecode.InstructionFactory;
import com.github.zxh.classpy.classfile.constant.ConstantPool;
import com.github.zxh.classpy.classfile.datatype.U4;
import com.github.zxh.classpy.classfile.reader.ClassReader;
import com.github.zxh.classpy.classfile.bytecode.Instruction;
import com.github.zxh.classpy.classfile.bytecode.Opcode;

class Code(val codeLength: U4) : ClassComponent() {
    
    override fun readContent(reader: ClassReader) {
        val startPosition = reader.position;
        val endPosition = startPosition + codeLength.value;
        
        while (reader.position < endPosition) {
            val pc = reader.position - startPosition;
            val b = reader.getByte(reader.position);
            val opcode = Opcode.XXX.valueOf(java.lang.Byte.toUnsignedInt(b));
            val instruction = InstructionFactory.create(opcode, pc);
            instruction.read(reader);
            add(instruction);
        }
    }

    override fun afterRead(cp: ConstantPool) {
        val instructions = super.subComponents!!;

        val maxPc = (instructions[instructions.size - 1] as Instruction).pc;
        val pcWidth = maxPc.toString().length;
        val fmtStr = "%0" + pcWidth + "d";
        for (c in instructions) {
            val instruction = c as Instruction;
            instruction.name = String.format(fmtStr, instruction.pc);
        }
    }

}
